### DISCUSSION TITLE
Run workflow only when the previous run is completed

### DISCUSSION THREAD

**Discussion by GORA-SAG** [community/community#55027] [Actions] [answered]
### Select Topic Area

Question

### Body

Hi Team,
I have a github workflow which will get triggered during on-merge. It has 3 jobs and i have only one runner. Right now manually i need to check whether there's no run in in-progress state, because if i trigger the workflow again then there is possibility that Jobs in Run 2 can pick the runner and job 1 will be in waiting state. Is there a way to stop, queue or disable workflow run if a workflow is already running.

**Comment by Yarden-zamir** [comment, marked as answer]
To solve this you can use [concurency groups](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#concurrency) either on the [job level](https://docs.github.com/en/actions/using-jobs/using-concurrency) or on the workflow level. For you, probably the latter, for example

```yaml
name: merge ci stuff
concurrency:
  group: ${{ github.workflow }}
```
which should only let workflows that fit this group run one at a time

you can also cancel runs from the same group when a new one is queued using `cancel-in-progress: true`

You can find examples for cuncurency groups [here](https://github.com/search?q=concurrency+path%3A.github%2Fworkflows%2F*.yaml&type=code&ref=advsearch)

**Reply by GORA-SAG** [reply]
Thanks a lot Yarden-zamir, this saved my day. Have a nice day.

**Reply by Yarden-zamir** [reply]
You are welcome!

**Reply by rpratama-codes** [reply]
cool you also save my day

**Reply by whyredfire** [reply]
_"When a concurrent job or workflow is queued, if another job or workflow using the same concurrency group in the repository is in progress, the queued job or workflow will be pending. Any previously pending job or workflow in the concurrency group will be canceled."_ (https://docs.github.com/en/actions/how-tos/write-workflows/choose-when-workflows-run/control-workflow-concurrency)

Using your example, if I trigger 5 workflow runs:

- workflow 1 (running) completes
- workflows 2–4 are canceled as newer workflows enter the queue
- workflow 5 (latest) runs

`cancel-in-progress: false` only prevents the currently running workflow from being canceled. GitHub will still cancel older pending workflows whenever a new one is queued.

In my case, every build matters and cannot be skipped. I need a true FIFO queue where all workflow runs execute sequentially without cancellation. Is this achievable with GitHub Actions, or is an external queuing system required?

**Reply by a-ShivamSrivastava** [reply]
I do have the same question. Every build matters and can not be skipped. Is this achievable with Github Actions or  is an external queuing system required?

**Comment by ashurauza** [comment]
You can prevent overlapping workflow runs by using GitHub Actions concurrency at the workflow level so only one run executes at a time. Add a concurrency group like this:

concurrency:
  group: ${{ github.workflow }}
  cancel-in-progress: false


This ensures that if another run is triggered while one is already running, it will be queued instead of starting immediately. If you instead want new runs to cancel previous ones automatically, set cancel-in-progress: true. Defining concurrency at the workflow level is ideal when you only have one runner and want the full pipeline to execute sequentially without manual checks.
